File: app/src/main/java/com/s95ammar/weeklyschedule/views/fragments/CategoryRefactorDialog.java
package com.s95ammar.weeklyschedule.views.fragments;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.s95ammar.weeklyschedule.R;
import com.s95ammar.weeklyschedule.models.Category;

import java.io.Serializable;
import java.util.HashSet;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.s95ammar.weeklyschedule.views.fragments.CategoriesListFragment.CategoriesListManager.KEY_CATEGORY;
import static com.s95ammar.weeklyschedule.views.fragments.CategoriesListFragment.CategoriesListManager.KEY_INDEX;

public class CategoryRefactorDialog extends AppCompatDialogFragment {
    private static final String TAG = "CategoryRefactorDialog";
    protected @BindView(R.id.text_add_category_preview_value) TextView mTextViewPreview;
    protected @BindView(R.id.eText_add_category_name) EditText mEditTextName;
    protected @BindView(R.id.view_add_category_fill_color) View mViewFillColor;
    protected @BindView(R.id.view_add_category_text_color) View mViewTextColor;
    private CategoryRefactor mListener;
    private static final String ADD_TITLE = "New Category";
    private static final String EDIT_TITLE = "Edit Category";

    public interface Action {
        int ADD = -1;
        // REMAKE = any value other than -1 (value will be used as index)
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof CategoryRefactor) {
            mListener = (CategoryRefactor) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement CategoryRefactor");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_add_category, null);
        Serializable obj = getArguments().getSerializable(KEY_CATEGORY);
        ButterKnife.bind(this, view);
        String title;
        Category category = null;
        if (obj instanceof Category) {
            category = (Category) obj;
            title = EDIT_TITLE;
        } else {
            title = ADD_TITLE;
        }
        setViews(category);
        final int i = getArguments().getInt(KEY_INDEX);
        return builder.setView(view)
                .setTitle(title)
                .setNegativeButton("Cancel", null)
                .setPositiveButton("Ok", onOkListener(i))
                .create();
    }

    private void setViews(@Nullable Category category) {
        mTextViewPreview.setText(R.string.category_preview_value);
        if (category == null) {
            mTextViewPreview.setBackgroundColor(((ColorDrawable) mViewFillColor.getBackground()).getColor());
        } else {
            mTextViewPreview.setTextColor(category.getTextColor());
            mTextViewPreview.setBackgroundColor(category.getFillColor());
            mEditTextName.setText(category.getName());
            mEditTextName.setSelection(mEditTextName.getText().length());
            mViewFillColor.setBackgroundColor(category.getFillColor());
            mViewTextColor.setBackgroundColor(category.getTextColor());
        }
    }

    private DialogInterface.OnClickListener onOkListener(final int i) {
        return (dialog, which) -> {
            String name = mEditTextName.getText().toString();
            if (!name.isEmpty()) {
                Category category = new Category(name,
                        ((ColorDrawable) mViewFillColor.getBackground()).getColor(),
                        ((ColorDrawable) mViewTextColor.getBackground()).getColor(),
                        new HashSet<>());
                mListener.applyCategory(category, i);
            } else {
                Toast.makeText(getActivity(), R.string.category_empty_name, Toast.LENGTH_SHORT).show();
            }
        };
    }

    @OnClick({R.id.view_add_category_fill_color, R.id.view_add_category_text_color})
    public void openColorPicker(View v) {
        switch (v.getId()) {
            case R.id.view_add_category_fill_color:
                mListener.openColorPicker(this, R.id.view_add_category_fill_color, ((ColorDrawable) v.getBackground()).getColor());
                break;
            case R.id.view_add_category_text_color:
                mListener.openColorPicker(this, R.id.view_add_category_text_color, ((ColorDrawable) v.getBackground()).getColor());
                break;
        }
    }

    public void receiveColor() {
        Bundle bundle = getArguments();
        int viewId = bundle.getInt(CategoryRefactor.KEY_ID);
        int color = bundle.getInt(CategoryRefactor.KEY_COLOR);
        switch (viewId) {
            case R.id.view_add_category_fill_color:
                mViewFillColor.setBackgroundColor(color);
                mTextViewPreview.setBackgroundColor(color);
                break;
            case R.id.view_add_category_text_color:
                mViewTextColor.setBackgroundColor(color);
                mTextViewPreview.setTextColor(color);
                break;
        }
    }

    public interface CategoryRefactor {
        void applyCategory(Category category, int i);
        void openColorPicker(CategoryRefactorDialog dialog, int viewId, int defaultColor);
        String KEY_ID = "id";
        String KEY_COLOR = "color";
    }
}

File: app/src/main/java/com/s95ammar/weeklyschedule/views/fragments/dialogs/CategoryRefactorDialog.kt
package com.s95ammar.weeklyschedule.views.fragments.dialogs

import android.app.Dialog
import android.content.DialogInterface
import android.graphics.PorterDuff
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorInt
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.core.graphics.drawable.DrawableCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.s95ammar.weeklyschedule.R
import com.s95ammar.weeklyschedule.models.data.Category
import com.s95ammar.weeklyschedule.util.*
import com.s95ammar.weeklyschedule.viewModels.CategoriesListViewModel
import dagger.android.support.DaggerDialogFragment
import kotlinx.android.synthetic.main.dialog_add_category.*
import javax.inject.Inject


class CategoryRefactorDialog : DaggerDialogFragment() {
	private var mode = Mode.ADD
	private lateinit var editedCategory: Category
	@StringRes private val addTitle = R.string.category_add_title
	private var categoryName = ""
		set(value) {
			field = value
			editText_refactor_category_name.setText(categoryName)
		}
	@ColorInt private var selectedFillColor = COLOR_GREEN
		set(value) {
			field = value
			assignFillColor()
		}
	@ColorInt private var selectedTextColor = COLOR_BLACK
		set(value) {
			field = value
			assignTextColor()
		}


	@Inject
	lateinit var factory: ViewModelProvider.Factory
	private lateinit var viewModel: CategoriesListViewModel
	private var dialogView: View? = null

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		dialogView = activity!!.layoutInflater.inflate(R.layout.dialog_add_category, null)
		return AlertDialog.Builder(activity!!)
				.setView(dialogView)
				.setTitle(addTitle)
				.setNegativeButton(R.string.cancel, null)
				.setPositiveButton(R.string.ok, onOkListener())
				.create()
	}

	// Workaround for "Can't access the Fragment View's LifecycleOwner before onCreateView()"
	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) = dialogView

	override fun onDestroy() {
		dialogView = null
		super.onDestroy()
	}

	override fun onActivityCreated(savedInstanceState: Bundle?) {
		super.onActivityCreated(savedInstanceState)
		activity?.let { viewModel = ViewModelProviders.of(it, factory).get(CategoriesListViewModel::class.java) }
		initObservers()
		setViews()
		setListeners()
	}

	private fun initObservers() {
		viewModel.editedCategory.observe(viewLifecycleOwner, Observer {
			it?.let { category ->
				editedCategory = category
				setModeEdit()
			}
		})
	}

	private fun setModeEdit() {
		mode = Mode.EDIT
		selectedFillColor = editedCategory.fillColor
		selectedTextColor = editedCategory.textColor
		categoryName = editedCategory.name
		dialog?.setTitle(R.string.category_edit_title)
	}

	private fun setViews() {
		editText_refactor_category_name.setText(categoryName)
		assignFillColor()
		assignTextColor()
	}

	private fun setListeners() {
		button_refactor_category_reverse_colors.setOnClickListener { reverseColors() }
		view_refactor_category_fill_color.setOnClickListener { onColorClicked(ColorDetails(selectedFillColor, ColorTarget.FILL)) }
		view_refactor_category_text_color.setOnClickListener { onColorClicked(ColorDetails(selectedTextColor, ColorTarget.TEXT)) }
	}

	private fun reverseColors() {
		val fillColor = selectedFillColor
		selectedFillColor = selectedTextColor
		selectedTextColor = fillColor
	}

	private fun onColorClicked(colorDetails: ColorDetails) {
		viewModel.showColorPickerDialog(colorDetails)
		observeColorSelection()
	}

	private fun observeColorSelection() {
		viewModel.onCategoryColorSelected.observe(viewLifecycleOwner, Observer { colorDetails ->
			colorDetails?.let {
				when (it.target) {
					ColorTarget.FILL -> selectedFillColor = it.color
					ColorTarget.TEXT -> selectedTextColor = it.color
				}
			}
		})
	}

	private fun onOkListener() = DialogInterface.OnClickListener { _,_ ->
		if (editText_refactor_category_name.input.isNotBlank()) {
			val category = Category(editText_refactor_category_name.input, selectedFillColor, selectedTextColor)
			when (mode) {
				Mode.ADD -> viewModel.insert(category)
				Mode.EDIT -> viewModel.update(category.apply { id = editedCategory.id })
			}
		} else {
			toast(activity, R.string.category_empty_name)
		}
	}

	private fun assignFillColor() {
		view_refactor_category_fill_color.background.setColorFilter(selectedFillColor, PorterDuff.Mode.SRC) // changes color but not shape
		text_refactor_category_preview_value.setBackgroundColor(selectedFillColor)
	}

	private fun assignTextColor() {
		view_refactor_category_text_color.background.setColorFilter(selectedTextColor, PorterDuff.Mode.SRC)
		text_refactor_category_preview_value.setTextColor(selectedTextColor)
	}

	override fun onDetach() {
		viewModel.clearRefactorDialogValues()
		super.onDetach()
	}
}
